src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp

```diff
@@ -334,16 +359,49 @@ void ReplicatedMergeTreeSink::commitPart(
 
     bool is_already_existing_part = false;
 
-    while (true)
+    /// for retries due to keeper error
+    bool part_committed_locally_but_zookeeper = false;
+    Coordination::Error write_part_info_keeper_error = Coordination::Error::ZOK;
+
+    ZooKeeperRetriesControl retries_ctl("commitPart", zookeeper_retries_info);
+    retries_ctl.retryLoop([&]()
     {
+        zookeeper->setKeeper(storage.getZooKeeper());
+
+        /// if we are in retry, check if last iteration was actually successful
+        /// we could get network error on latest keeper operation in iteration
+        /// but operation could be completed by keeper server
+        if (retries_ctl.isRetry())
+        {
+            if (part_committed_locally_but_zookeeper)
```

Comment:
/// If this flag is true, then part is in Active state, and we'll not retry anymore
/// we only check if part was committed to zk and return success or failure correspondingly
/// Note: if commit to zk failed, then cleanup thread will mark the part as Outdated later